File: either-java17/src/test/java/dev/andrybak/util/function/java17/EitherTest.java
// SPDX-License-Identifier: MIT
package dev.andrybak.util.function.java17;

import org.junit.jupiter.api.Test;

import java.util.Collection;
import java.util.Collections;
import java.util.List;
import java.util.function.Function;

import static org.junit.jupiter.api.Assertions.assertAll;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertSame;
import static org.junit.jupiter.api.Assertions.fail;

public class EitherTest {
	@Test
	void testThatLeftCanBeConstructed() {
		Either<String, Integer> leftValue = Either.left("Left value");
		assertNotNull(leftValue);
		assertEquals(Left.class, leftValue.getClass());
	}

	@Test
	void testThatRightCanBeConstructed() {
		Either<String, Integer> rightValue = Either.right(42);
		assertNotNull(rightValue);
		assertEquals(Right.class, rightValue.getClass());
	}

	@Test
	void testThatLeftCanBeMatched() {
		Either<String, Integer> leftValue = Either.left("bar");
		assertEquals("foobar", leftValue.match(
				a -> "foo" + a,
				b -> "Right value " + b
		));
	}

	@Test
	void testThatRightCanBeMatched() {
		Either<String, Integer> rightValue = Either.right(42);
		assertEquals("Right value 42", rightValue.match(
				a -> "foo" + a,
				b -> "Right value " + b
		));
	}

	@Test
	void testThatMatchAcceptsFunctionWithSuperClassInput() {
		assertAll(() -> {
			Either<String, Integer> leftValue = Either.left("bar");
			Function<CharSequence, String> f = cs -> "foo" + cs.toString();
			assertEquals("foobar", leftValue.match(
					f,
					b -> "Right value " + b
			));
		}, () -> {
			Either<String, Integer> rightValue = Either.right(42);
			Function<Number, String> g = n -> n.toString() + "bar";
			assertEquals("42bar", rightValue.match(
					a -> "Left value " + a,
					g
			));
		});
	}

	@Test
	void testThatMatchAcceptsFunctionWithSubClassOutput() {
		// dummy is used to force the return type to be `Collection<String>`
		Collection<String> dummy = Collections.emptySet();
		assertAll(() -> {
			Either<String, Integer> leftValue = Either.left("bar");
			Function<String, List<String>> f = s -> List.of("foo", s);
			assertEquals(List.of("foo", "bar"), leftValue.match(
					f,
					(Integer b) -> dummy
			));
		}, () -> {
			Either<String, Integer> rightValue = Either.right(42);
			Function<Integer, List<String>> g = i -> List.of(i.toString(), "bar");
			assertEquals(List.of("42", "bar"), rightValue.match(
					(String a) -> dummy,
					g
			));
		});
	}

	@Test
	void testThatEitherAcceptsFunctionWithSuperClassInput() {
		assertAll("either(f,g)", () -> {
			Either<String, Integer> leftValue = Either.left("bar");
			Function<CharSequence, String> takesSuperclass = cs -> "foo" + cs.toString();
			assertEquals(
					"foobar",
					Either.either(takesSuperclass, b -> "Right value " + b).apply(leftValue)
			);
		}, () -> {
			Either<String, Integer> rightValue = Either.right(42);
			Function<Number, String> takesSuperClass = n -> n.toString() + "bar";
			assertEquals(
					"42bar",
					Either.either(a -> "Left value " + a, takesSuperClass).apply(rightValue)
			);
		});
		assertAll("either(f,g,e)", () -> {
			Either<String, Integer> leftValue = Either.left("bar");
			Function<CharSequence, String> takesSuperclass = cs -> "foo" + cs.toString();
			assertEquals(
					"foobar",
					Either.either(takesSuperclass, b -> "Right value " + b, leftValue)
			);
		}, () -> {
			Either<String, Integer> rightValue = Either.right(42);
			Function<Number, String> takesSuperClass = n -> n.toString() + "bar";
			assertEquals(
					"42bar",
					Either.either(a -> "Left value " + a, takesSuperClass, rightValue)
			);
		});
	}

	@Test
	void testThatEitherAcceptsEitherWithSubClass() {
		class SuperClass {
			@Override
			public String toString() {
				return "SuperClass";
			}
		}
		class SubClass extends SuperClass {
			@Override
			public String toString() {
				return "SubClass";
			}
		}
		assertAll("either(f,g,e) with forced type parameters", () -> {
			Either<SubClass, Integer> leftValue = Either.left(new SubClass());
			Function<SuperClass, String> takesSuperclass = superObj -> "foo" + superObj.toString();
			Function<Integer, String> g = (Integer b) -> "Right value " + b;
			assertEquals(
					"fooSubClass",
					Either.<SuperClass, Integer, String>either(takesSuperclass, g, leftValue)
			);
		}, () -> {
			Either<Integer, SubClass> rightValue = Either.right(new SubClass());
			Function<Integer, String> f = (Integer b) -> "Left value " + b;
			Function<SuperClass, String> takesSuperclass = superObj -> "foo" + superObj.toString();
			assertEquals(
					"fooSubClass",
					Either.<Integer, SuperClass, String>either(f, takesSuperclass, rightValue)
			);
		});
	}

	@Test
	void testThatEitherAcceptsFunctionWithSubClassOutput() {
		// dummy is used to force the return type to be `Collection<String>`
		Collection<String> dummy = Collections.emptySet();
		assertAll("either(f,g)", () -> {
			Either<String, Integer> leftValue = Either.left("bar");
			Function<String, List<String>> f = s -> List.of("foo", s);
			assertEquals(List.of("foo", "bar"), Either.either(
					f,
					(Integer b) -> dummy
			).apply(leftValue));
		}, () -> {
			Either<String, Integer> rightValue = Either.right(42);
			Function<Integer, List<String>> g = i -> List.of(i.toString(), "bar");
			assertEquals(List.of("42", "bar"), Either.either(
					(String a) -> dummy,
					g
			).apply(rightValue));
		});
		assertAll("either(f,g,e)", () -> {
			Either<String, Integer> leftValue = Either.left("bar");
			Function<String, List<String>> f = s -> List.of("foo", s);
			assertEquals(List.of("foo", "bar"), Either.either(
					f,
					(Integer b) -> dummy,
					leftValue
			));
		}, () -> {
			Either<String, Integer> rightValue = Either.right(42);
			Function<Integer, List<String>> g = i -> List.of(i.toString(), "bar");
			assertEquals(List.of("42", "bar"), Either.either(
					(String a) -> dummy,
					g,
					rightValue
			));
		});
	}

	@Test
	void testAcceptOnLeft() {
		Either<String, Integer> leftValue = Either.left("foo");
		final int[] leftCount = {0};
		leftValue.accept(
				a -> {
					leftCount[0]++;
					assertEquals("foo", a);
				},
				b -> fail("unexpected Right")
		);
		assertEquals(1, leftCount[0]);
	}

	@Test
	void testAcceptOnRight() {
		Either<String, Integer> rightValue = Either.right(42);
		final int[] rightCount = {0};
		rightValue.accept(
				a -> fail("unexpected Left"),
				b -> {
					rightCount[0]++;
					assertEquals(42, b);
				}
		);
		assertEquals(1, rightCount[0]);
	}

	@Test
	void testPeekOnLeft() {
		Either<String, Integer> leftValue = Either.left("foo");
		final int[] leftCount = {0};
		Either<String, Integer> res = leftValue.peek(
				a -> {
					leftCount[0]++;
					assertEquals("foo", a);
				},
				b -> fail("unexpected Right")
		);
		assertEquals(1, leftCount[0]);
		assertSame(leftValue, res);
	}

	@Test
	void testPeekOnRight() {
		Either<String, Integer> rightValue = Either.right(42);
		final int[] rightCount = {0};
		Either<String, Integer> res = rightValue.peek(
				a -> fail("unexpected Left"),
				b -> {
					rightCount[0]++;
					assertEquals(42, b);
				}
		);
		assertSame(rightValue, res);
		assertEquals(1, rightCount[0]);
	}
}

File: either-kotlin/src/test/kotlin/dev/andrybak/util/function/EitherTest.kt
package dev.andrybak.util.function

import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertInstanceOf
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertAll

internal class EitherTest {
	@Test
	internal fun testThatLeftCanBeConstructed() {
		val leftValue: Either<String, Int> = Either.left("Hello, world!")
		assertNotNull(leftValue)
		assertEquals(Either.Left::class, leftValue::class)
	}

	@Test
	internal fun testThatRightCanBeConstructed() {
		val rightValue: Either<String, Int> = Either.right(42)
		assertNotNull(rightValue)
		assertEquals(Either.Right::class, rightValue::class)
	}

	@Test
	internal fun testThatLeftCanBeMatchedByWhen() {
		val leftValue: Either<String, Int> = Either.left("bar")
		assertEquals(
			"foobar",
			when (leftValue) {
				is Either.Left -> "foo" + leftValue.leftValue
				is Either.Right -> "Right value " + leftValue.rightValue
			}
		)
	}

	@Test
	internal fun testThatLeftCanBeMatchedByMatch() {
		val leftValue: Either<String, Int> = Either.left("bar")
		val f: (String) -> String = { a -> "foo$a" }
		val g: (Int) -> String = { b -> "Right value $b" }
		assertEquals("foobar", leftValue.match(f, g))
	}

	@Test
	internal fun testThatRightCanBeMatchedByMatch() {
		val rightValue: Either<String, Int> = Either.right(42)
		val f: (String) -> String = { a -> "foo$a" }
		val g: (Int) -> String = { b -> "Right value $b" }
		assertEquals("Right value 42", rightValue.match(f, g))
	}

	@Test
	internal fun testThatLeftCanBeMatchedByMatchWithMethodRef() {
		val leftValue: Either<String, Int> = Either.left("bar")
		assertEquals("foobar", leftValue.match(::foo, ::bar))
	}

	@Test
	internal fun testThatRightCanBeMatchedByMatchWithMethodRef() {
		val rightValue: Either<String, Int> = Either.right(42)
		assertEquals("Right value 42", rightValue.match(::foo, ::bar))
	}

	@Test
	internal fun testThatRightCanBeMatchedByWhen() {
		val rightValue: Either<String, Int> = Either.right(42)
		assertEquals(
			"Right value 42",
			when (rightValue) {
				is Either.Left -> "foo" + rightValue.leftValue
				is Either.Right -> "Right value " + rightValue.rightValue
			}
		)
	}

	@Test
	internal fun testThatLeftCanBeMatchedByEitherFunction() {
		val leftValue: Either<String, Int> = Either.left("bar")
		assertEquals("foobar", either({ a -> "foo$a" }, { b -> "Right value $b" }, leftValue))
	}

	@Test
	internal fun testThatRightCanBeMatchedByEitherFunction() {
		val rightValue: Either<String, Int> = Either.right(42)
		assertEquals("Right value 42", either({ a -> "foo$a" }, { b -> "Right value $b" }, rightValue))
	}

	@Test
	internal fun testThatLeftCanBeMatchedByEitherFunction2() {
		val leftValue: Either<String, Int> = Either.left("bar")
		val f: (Either<String, Int>) -> String = either({ a -> "foo$a" }, { b -> "Right value $b" })
		assertEquals("foobar", f(leftValue))
	}

	@Test
	internal fun testThatRightCanBeMatchedByEitherFunction2() {
		val rightValue: Either<String, Int> = Either.right(42)
		val f: (Either<String, Int>) -> String = either({ a -> "foo$a" }, { b -> "Right value $b" })
		assertEquals("Right value 42", f(rightValue))
	}

	@Test
	internal fun testThatEitherOfSubClassesCanBeUsedAsParameter() {
		val f: (Either<CharSequence, Number>) -> Unit = { e ->
			when (e) {
				is Either.Left -> assertInstanceOf(CharSequence::class.java, e.leftValue)
				is Either.Right -> assertInstanceOf(Number::class.java, e.rightValue)
			}
		}
		val leftValue: Either<String, Int> = Either.left("foo")
		val rightValue: Either<String, Int> = Either.right(42)
		f.invoke(leftValue)
		f.invoke(rightValue)
	}

	@Test
	internal fun testThatEitherAcceptsFunctionWithSuperClassInput() {
		val f: (CharSequence) -> String = { cs -> "foo$cs" }
		val g: (Number) -> String = { n -> "${n}bar" }
		val leftValue: Either<String, Int> = Either.left("bar")
		assertEquals("foobar", either(f, g, leftValue))
		assertEquals("foobar", either(f, g).invoke(leftValue))
		val rightValue: Either<String, Int> = Either.right(42)
		assertEquals("42bar", either(f, g, rightValue))
		assertEquals("42bar", either(f, g).invoke(rightValue))
	}

	@Test
	internal fun testThatEitherAcceptsFunctionWithSubClassOutput() {
		assertAll({
			val f: (String) -> List<String> = { cs -> listOf("foo", cs) }
			val dummy: (Int) -> Collection<String> = { _ -> emptySet() }
			val leftValue: Either<String, Int> = Either.left("bar")
			assertEquals(listOf("foo", "bar"), either(f, dummy, leftValue))
			assertEquals(listOf("foo", "bar"), either(f, dummy).invoke(leftValue))
		}, {
			val dummy: (String) -> Collection<String> = { _ -> emptySet() }
			val g: (Int) -> List<String> = { n -> listOf(n.toString(), "bar") }
			val rightValue: Either<String, Int> = Either.right(42)
			assertEquals(listOf("42", "bar"), either(dummy, g).invoke(rightValue))
			assertEquals(listOf("42", "bar"), either(dummy, g, rightValue))
		})
	}

	@Test
	fun testAcceptOnLeft() {
		val leftValue: Either<String, Int> = Either.left("foo")
		var leftCount = 0
		leftValue.accept(
			{ a: String ->
				leftCount++
				assertEquals("foo", a)
			},
			{ _: Int -> Assertions.fail<Any>("unexpected Right") }
		)
		assertEquals(1, leftCount)
	}

	@Test
	fun testAcceptOnRight() {
		val rightValue: Either<String, Int> = Either.right(42)
		var rightCount = 0
		rightValue.accept(
			{ _: String ->
				Assertions.fail<Any>(
					"unexpected Left"
				)
			},
			{ b: Int? ->
				rightCount++
				assertEquals(42, b)
			}
		)
		assertEquals(1, rightCount)
	}

	@Test
	fun testPeekOnLeft() {
		val leftValue: Either<String, Int> = Either.left("foo")
		var leftCount = 0
		val res: Either<String, Int> = leftValue.peek(
			{ a: String ->
				leftCount++
				assertEquals("foo", a)
			},
			{ _: Int ->
				Assertions.fail("unexpected Right")
			}
		)
		assertEquals(1, leftCount)
		Assertions.assertSame(leftValue, res)
	}

	@Test
	fun testPeekOnRight() {
		val rightValue: Either<String, Int> = Either.right(42)
		var rightCount = 0
		val res: Either<String, Int> = rightValue.peek(
			{ _: String ->
				Assertions.fail("unexpected Left")
			},
			{ b: Int ->
				rightCount++
				assertEquals(42, b)
			}
		)
		Assertions.assertSame(rightValue, res)
		assertEquals(1, rightCount)
	}

	companion object {
		fun foo(s: String): String = "foo$s"
		private fun bar(i: Int): String = "Right value $i"
	}
}
